lean4-infoview/src/infoview/info.tsx

```diff
@@ -307,8 +318,8 @@ function InfoAux(props: InfoProps) {
         let allReq
         if (sv?.hasWidgetsV1()) {
             // Start both goal requests before awaiting them.
-            const goalsReq = getInteractiveGoals(rs, pos);
-            const termGoalReq = getInteractiveTermGoal(rs, pos);
+            const goalsReq = getInteractiveGoals(rpcSess0, DocumentPosition.toTdpp(pos));
```

Comment:
I think if we start bundling positions with `RpcSession`s, we should also make it possible to retrieve the position from there and drop the `pos` variable here. It would certainly help avoid consistency issues.

Oh, and let's call the type `RpcSessionAtPos` for clarity.

Reply:
> Oh, and let's call the type RpcSessionAtPos for clarity.

Ha, I originally used exactly that name, but then I changed it back to `RpcSession` because it's shorter. :smile:  I've changed it back to `RpcSessionAtPos`.

> possible to retrieve the position from there and drop the `pos` variable here.

That's exactly the reason why I *don't* want to make the position available.  The general contract with this API is that you should never mix references from two different `RpcSession`s.  So if you want to show goals from two different positions in a widget, then you'd call `getInteractiveGoal` twice on the same `RpcSession` but with a different position argument.  This doesn't work if e.g. the goal widget takes the position from the session.  The only purpose of the position argument is to pick up `serverRpcMethod` declarations in the current file, I don't want to overload it.

Another thing: widgets get an implicit `RpcSession` via a React context.  Right now they can just use this `RpcSession` to make queries at any position, and `<InteractiveCode>` and co. will automatically pick up the right session.  If e.g. the goal request took the position from the `RpcSession`, then you'd be back to manually juggling with `RpcSession`s.

Reply:
> So if you want to show goals from two different positions in a widget, then you'd call `getInteractiveGoal` twice on the same `RpcSession` but with a different position argument.

Wait what, really? I expected to use the api like `getGoal(rpcSessionForFile.at(pos0))` and `getGoal(rpcSessionForFile.at(pos1))`, especially since the get goal method at two different positions might in principle be different and require a different RPC context.

Reply:
Yes, really.  The API doesn't expose `RpcSessionForFile` either, everything's hidden behind `RpcSessionAtPos`.  (We could certainly add some escape hatches though.)  But your average widget should never need two different sessions.  In the example, the two goals are probably in the same command so it doesn't matter for `serverRpcMethod` which position you choose.

> require a different RPC context.

They only require a different RPC context if you want different `serverRpcMethod`s available.  (Please correct me if there's any other difference.  But that was the original reason why we had a position and not just the file.)  And if you have two context you need to decide for every RPC call the session it should go to.

Reply:
I'm a bit confused by the discussion here. The whole concern of making requests at two different positions using a single `RpcSessionAtPos` is theoretical since we don't actually do that anywehre, right? But while in theoretical-API-usage land, it seems wrong to just hope that the two positions have the same `serverRpcMethod`s, in particular because it would be relying on an implementation detail of the two `RpcSessionAtPos`, namely the fact that the two positions share a single underlying session.

Now I am thinking I was conceptualizing `RpcSessionAtPos` incorrectly. How about we make the contract explicit. We'd say that `RpcSessionAtPos` gives you access to all the `serverRpcMethods` available at its pos, and that it's safe to make calls with it at that position or *anywhere after*, but not before. In other words, an `RpcSessionAtPos` *is* morally a fixed set of `serverRpcMethods` together with the usual ref management state.

Reply:
> The whole concern of making requests at two different positions using a single `RpcSessionAtPos` is theoretical

At this moment you're completely right.  I think this came up in the widget discussion somewhere.

> it's safe to make calls with it at that position or anywhere after, but not before.

You probably have the better picture here, why is it unsafe to make a call to e.g. `getInteractiveGoals` at a position before?

> RpcSessionAtPos gives you access to all the serverRpcMethods available at its pos [...] RpcSessionAtPos is morally a fixed set of serverRpcMethods together with the usual ref management state.

I'll put something like this in the docstring.

Reply:
> You probably have the better picture here, why is it unsafe to make a call to e.g. `getInteractiveGoals` at a position before?

What I had in mind is just that RPC methods exist from the registration point onwards. In fact because the `sessionPos` is fixed, it might even be possible to make a call with `pos < sessionPos`, and it would still find the right RPC method since it is looked up using `sessionPos`. However if the RPC method relies on some objects which are registered at `objPos` with `pos < objPos < sessionPos`, I have no idea what would happen. Anyway this is unlikely to come up in actual code.